pkg/middleware/rbac.go

```diff
@@ -0,0 +1,103 @@
+package middleware
+
+import (
+	"strings"
+	"time"
+
+	"github.com/content-services/content-sources-backend/pkg/client"
+	"github.com/labstack/echo/v4"
+	echo_middleware "github.com/labstack/echo/v4/middleware"
+)
+
+// This middleware will add rbac feature to the service
+
+// https://echo.labstack.com/cookbook/middleware/
+// https://github.com/labstack/echo/tree/master/middleware
+
+const clientTimeout = 10 * time.Second
+const xrhidHeader = "X-Rh-Identity"
+
+type Rbac struct {
+	BaseUrl string
+	Skipper echo_middleware.Skipper
+
+	client client.Rbac
+}
+
+func NewRbac(config Rbac, proxy client.Rbac) echo.MiddlewareFunc {
+	config.client = proxy
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper != nil && config.Skipper(c) {
+				return next(c)
+			}
+
+			resource := fromPathToResource(c.Request().URL.Path)
+			if resource == "" {
+				return echo.ErrUnauthorized
+			}
+
+			verb := fromHttpVerbToRbacVerb(c.Request().Method)
+			if verb == client.RbacVerbUndefined {
+				return echo.ErrUnauthorized
+			}
+
+			xrhid := c.Request().Header.Get(xrhidHeader)
+			if xrhid == "" {
+				return echo.ErrUnauthorized
+			}
+
+			allowed, err := config.client.Allowed(xrhid, resource, verb)
+			if err != nil {
+				return echo.ErrInternalServerError
+			}
+			if !allowed {
+				return echo.ErrUnauthorized
+			}
+
+			return echo.ErrUnauthorized
+		}
+	}
+}
+
+//
+// Private functions
+//
+
+func fromHttpVerbToRbacVerb(httpMethod string) client.RbacVerb {
+	switch httpMethod {
+	case echo.GET:
+		return client.RbacVerbRead
+	case echo.POST:
+		return client.RbacVerbWrite
+	case echo.PATCH:
+		return client.RbacVerbWrite
+	case echo.PUT:
+		return client.RbacVerbWrite
+	case echo.DELETE:
+		return client.RbacVerbWrite
+
+	default:
+		return client.RbacVerbUndefined
+	}
+}
```

Comment:
i think our routes don't fully match this, for example:
```
    "method": "POST",
    "path": "/api/content-sources/v1/rpms/names",
```
would be associated with a Read permission.  It'd be nice if we had some method that could register the route alongside the permission:

```
	engine.GET("/repositories/", rh.listRepositories, rbac.RbacVerbRead)
	engine.GET("/repositories/:uuid", rh.fetch, rbac.RbacVerbRead)
	engine.PUT("/repositories/:uuid", rh.fullUpdate, rbac.RbacVerbWrite)
        etc...
```

Reply:
thanks! inspiring! I have a solution in mind that would works!

Reply:
I have added a mechanism to map the api requests to the permissions.
- I create a new type which is a map[string]map[string]Permission.
- First map map the request method.
- Second map map the api path once its prefix is removed.
- The value stored is a struct for the permission name and verb for that http operation.

By default, if no mapping is found, it is rejected with unauthorized.

If a mapping is found, then the rbac client is called with the permission to check for the operation, and reject or continuous with the operation depending on the result.